pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
+    data: DataFrame
+        timeseries data from the BSRN archive, see
+        :func:`pvlib.iotools.read_bsrn` for fields.
+    metadata: dict
+        Dictionary containing metadata (primarily from LR0004).
+
+    See Also
+    --------
+    pvlib.iotools.read_bsrn, pvlib.iotools.get_bsrn
+
+    """
+    # Parse metadata
+    fbuf.readline()  # first line should be *U0001, so read it and discard
+    date_line = fbuf.readline()  # second line contains important metadata
+    start_date = pd.Timestamp(year=int(date_line[7:11]),
+                              month=int(date_line[3:6]), day=1,
+                              tz='UTC')  # BSRN timestamps are UTC
+
+    metadata = {}  # Initilize dictionary containing metadata
+    metadata['start date'] = start_date
+    metadata['station identification number'] = int(date_line[:3])
+    metadata['version of data'] = int(date_line.split()[-1])
+    for line in fbuf:
+        if line[2:6] == '0004':  # stop once LR0004 has been reached
+            break
+        elif line == '':
+            raise ValueError('Mandatory record LR0004 not found.')
+    metadata['date when station description changed'] = fbuf.readline().strip()
+    metadata['surface type'] = int(fbuf.readline(3))
+    metadata['topography type'] = int(fbuf.readline())
+    metadata['address'] = fbuf.readline().strip().strip()
+    metadata['telephone no. of station'] = fbuf.readline(20).strip()
+    metadata['FAX no. of station'] = fbuf.readline().strip()
+    metadata['TCP/IP no. of station'] = fbuf.readline(15).strip()
+    metadata['e-mail address of station'] = fbuf.readline().strip()
+    metadata['latitude'] = float(fbuf.readline(8))
+    metadata['longitude'] = float(fbuf.readline(8))
+    metadata['altitude'] = int(fbuf.readline(5))
+    metadata['identification of "SYNOP" station'] = fbuf.readline().strip()
+    metadata['date when horizon changed'] = fbuf.readline().strip()
+    # Pass last section of LR0004 containing the horizon elevation data
+    horizon = []  # list for raw horizon elevation data
+    while True:
+        line = fbuf.readline()
+        if ('*' in line) | (line == ''):
+            break
+        else:
+            horizon += [int(i) for i in line.split()]
+    horizon = pd.Series(horizon[1::2], horizon[::2], name='horizon_elevation',
+                        dtype=int).drop(-1, errors='ignore').sort_index()
+    horizon.index.name = 'azimuth'
+    metadata['horizon'] = horizon
+
+    # Read file and store the starting line number and number of lines for
+    # each logical record (LR)
+    fbuf.seek(0)  # reset buffer to start of file
+    lr_startrow = {}  # Dictionary of starting line number for each LR
+    lr_nrows = {}  # Dictionary of end line number for each LR
+    for num, line in enumerate(fbuf):
+        if line.startswith('*'):  # Find start of all logical records
+            if len(lr_startrow) >= 1:
+                lr_nrows[lr] = num - max(lr_startrow.values())-1  # noqa: F821

```

Comment:
```suggestion
                lr_nrows[lr] = num - lr_startrow[lr]
```
maybe clearer?  might need a -1 or something too

Reply:
Nice tough. Certainly makes it much clearer. The -1 is needed fyi.